src/error.rs

```diff
@@ -1,26 +1,65 @@
+use thiserror::Error;
+
+use crate::proto::ConnectError;
+
+/// The set of observable errors when interacting with a Faktory server.
+#[derive(Debug, Error)]
+#[allow(clippy::manual_non_exhaustive)]
+pub enum Error {
+    /// The connection to the server, or one of its prerequisites, failed.
+    #[error("connection error: {0}")]
+    Connect(#[from] ConnectError),
+
+    /// Underlying io layer errors.
+    /// These are overwhelmingly network communication errors on the socket connection to the server.
+    #[error("underlying i/o: {0}")]
+    GenericIO(#[from] std::io::Error),
+
+    /// Application-level errors.
+    /// These generally indicate a mismatch between what the client expects and what the server expects.
+    #[error("protocol: {0}")]
+    Protocol(#[from] ProtocolError),
+
+    /// Faktory payloads are JSON encoded.
+    /// This error is one that was encountered when attempting to deserialize a response from the server.
+    /// These generally indicate a mismatch between what the client expects and what the server provided.
+    #[error("deserialize payload: {0}")]
+    DeserializePayload(#[from] serde_json::Error),
+
+    /// Indicates an error in the underlying TLS stream.
+    #[cfg(feature = "tls")]
+    #[error("underlying tls stream: {0}")]
+    TlsStream(#[from] native_tls::Error),
+
+    // We're going to add more error types in the future
+    // https://github.com/rust-lang/rust/issues/44109
+    //
+    // This forces users to write pattern matches with a catch-all `_` arm.
+    #[error("unreachable")]
+    #[doc(hidden)]
+    __Nonexhaustive,
```

Comment:
I think we can safely adopt `#[non_exhaustive]` now, unless you think there's a strong reason not to?